services/brig/src/Brig/Run.hs

```diff
@@ -103,3 +124,51 @@ lookupRequestIdMiddleware :: (RequestId -> Wai.Application) -> Wai.Application
 lookupRequestIdMiddleware mkapp req cont = do
   let reqid = maybe def RequestId $ lookupRequestId req
   mkapp reqid req cont
+
+cleanExpiredPendingInvitations :: AppIO ()
+cleanExpiredPendingInvitations = do
+  safeForever "cleanExpiredPendingInvitations" $ do
+    today <- utctDay <$> (liftIO =<< view currentTime)
+    for_ [0 .. 9] $ \i ->
+      cleanUpDay (addDays (- i) today)
+    let d :: Int = 24 * 60 * 60
+    randomSecs <- liftIO (round <$> randomRIO @Double (0.5 * fromIntegral d, fromIntegral d))
+    threadDelay (randomSecs * 1_000_000)
+  where
+    cleanUpDay :: Day -> AppIO ()
+    cleanUpDay day = do
+      expiredEntries <- forTrackedExpirations day $ \exps ->
+        catMaybes
+          <$> ( for exps $ \(PendingActivationExpiration uid expiresAt tid) -> do
+                  isExpired <- (expiresAt <=) <$> (liftIO =<< view currentTime)
+                  if isExpired
+                    then do
+                      isPendingInvitation <- (Just PendingInvitation ==) <$> lookupStatus uid
+                      invExpired <- isNothing <$> Data.lookupInvitation tid (coerce uid)
+                      when (isPendingInvitation && invExpired) $ do
+                        API.deleteUserNoVerify uid
+                      pure (Just uid)
+                    else pure Nothing
+              )
+      unless (null expiredEntries) $
+        removeTrackedExpiration day expiredEntries
+
+    forTrackedExpirations :: Day -> ([PendingActivationExpiration] -> AppIO [UserId]) -> AppIO [UserId]
+    forTrackedExpirations day f = do
+      page <- searchTrackedExpirations day
+      go [] page
+      where
+        go :: [UserId] -> Page PendingActivationExpiration -> AppIO [UserId]
+        go users (Page hasMore result nextPage) = do
+          users' <- (<> users) <$> f result
+          if hasMore
+            then go users' =<< liftClient nextPage
+            else pure users'
+
+    safeForever :: (MonadIO m, MonadLogger m, MonadCatch m) => String -> m () -> m ()
+    safeForever funName action =
```

Comment:
I think this is duplicating existing code.  Can we move the old function somewhere from where we can call it here as well?  Or is it awkward to generalize?  (Not that important.)

Reply:
It's already duplicated 3 times. I created a separate issue for this https://wearezeta.atlassian.net/browse/BE-486